File: packages/sdk/java/packages/core/src/main/java/dev/edgebase/sdk/core/TableRef.java
// EdgeBase Java SDK — Collection reference & document reference.
// Immutable query builder pattern with full CRUD, batch operations, and database-live.
//
// All HTTP calls delegate to GeneratedDbApi (generated core).
// No hardcoded API paths — the core is the single source of truth.
package dev.edgebase.sdk.core;

import com.google.gson.Gson;
import dev.edgebase.sdk.core.generated.GeneratedDbApi;

import java.util.*;

/**
 * Immutable table reference with query builder.
 *
 * <p>
 * All chaining methods return a new instance — safe for reference sharing.
 *
 * <p>
 * Usage:
 *
 * <pre>{@code
 * ListResult posts = client.db("shared").table("posts")
 *         .where("status", "==", "published")
 *         .orderBy("createdAt", "desc")
 *         .limit(20)
 *         .getList();
 * }</pre>
 */
public class TableRef {
    private static final Gson gson = new Gson();

    private static String buildDatabaseLiveChannel(String namespace, String instanceId, String table, String docId) {
        String base = instanceId != null
                ? "dblive:" + namespace + ":" + instanceId + ":" + table
                : "dblive:" + namespace + ":" + table;
        return docId != null ? base + ":" + docId : base;
    }

    private final GeneratedDbApi core;
    private final String name;
    private final String namespace;
    private final String instanceId;
    private final DatabaseLiveClient databaseLive;
    private final List<FilterTuple> filters;
    private final List<FilterTuple> orFilters;
    private final List<String[]> sorts;
    private final Integer limitValue;
    private final Integer offsetValue;
    private final Integer pageValue;
    private final String searchValue;
    private final String afterCursor;
    private final String beforeCursor;

    public TableRef(GeneratedDbApi core, String name, String namespace, String instanceId, DatabaseLiveClient databaseLive) {
        this(core, name, namespace, instanceId, databaseLive,
                Collections.emptyList(), Collections.emptyList(), Collections.emptyList(),
                null, null, null, null, null, null);
    }

    private TableRef(GeneratedDbApi core, String name, String namespace, String instanceId, DatabaseLiveClient databaseLive,
            List<FilterTuple> filters, List<FilterTuple> orFilters, List<String[]> sorts,
            Integer limitValue, Integer offsetValue, Integer pageValue,
            String searchValue,
            String afterCursor, String beforeCursor) {
        this.core = core;
        this.name = name;
        this.namespace = namespace;
        this.instanceId = instanceId;
        this.databaseLive = databaseLive;
        this.filters = filters;
        this.orFilters = orFilters;
        this.sorts = sorts;
        this.limitValue = limitValue;
        this.offsetValue = offsetValue;
        this.pageValue = pageValue;
        this.searchValue = searchValue;
        this.afterCursor = afterCursor;
        this.beforeCursor = beforeCursor;
    }

    public String getName() {
        return name;
    }

    // ─── Core dispatch helpers (static vs dynamic DB) ───

    /** Dispatch a GET-style read to the correct generated core method. */
    @SuppressWarnings("unchecked")
    private Object coreGet(String method, Map<String, String> query, String id) {
        if (instanceId != null) {
            // Dynamic DB
            return switch (method) {
                case "list" -> core.dbListRecords(namespace, instanceId, name, query);
                case "get" -> core.dbGetRecord(namespace, instanceId, name, id, query);
                case "count" -> core.dbCountRecords(namespace, instanceId, name, query);
                case "search" -> core.dbSearchRecords(namespace, instanceId, name, query);
                default -> throw new IllegalArgumentException("Unknown method: " + method);
            };
        }
        // Single-instance DB
        return switch (method) {
            case "list" -> core.dbSingleListRecords(namespace, name, query);
            case "get" -> core.dbSingleGetRecord(namespace, name, id, query);
            case "count" -> core.dbSingleCountRecords(namespace, name, query);
            case "search" -> core.dbSingleSearchRecords(namespace, name, query);
            default -> throw new IllegalArgumentException("Unknown method: " + method);
        };
    }

    private Object coreInsert(Map<String, ?> body, Map<String, String> query) {
        if (instanceId != null) {
            return core.dbInsertRecord(namespace, instanceId, name, body, query);
        }
        return core.dbSingleInsertRecord(namespace, name, body, query);
    }

    private Object coreUpdate(String id, Map<String, ?> body) {
        if (instanceId != null) {
            return core.dbUpdateRecord(namespace, instanceId, name, id, body);
        }
        return core.dbSingleUpdateRecord(namespace, name, id, body);
    }

    private Object coreDelete(String id) {
        if (instanceId != null) {
            return core.dbDeleteRecord(namespace, instanceId, name, id);
        }
        return core.dbSingleDeleteRecord(namespace, name, id);
    }

    private Object coreBatch(Map<String, ?> body, Map<String, String> query) {
        if (instanceId != null) {
            return core.dbBatchRecords(namespace, instanceId, name, body, query);
        }
        return core.dbSingleBatchRecords(namespace, name, body, query);
    }

    private Object coreBatchByFilter(Map<String, ?> body, Map<String, String> query) {
        if (instanceId != null) {
            return core.dbBatchByFilter(namespace, instanceId, name, body, query);
        }
        return core.dbSingleBatchByFilter(namespace, name, body, query);
    }

    // ─── Query Builder (immutable — returns new instances) ───

    public TableRef where(String field, String op, Object value) {
        List<FilterTuple> newFilters = new ArrayList<>(filters);
        newFilters.add(new FilterTuple(field, op, value));
        return cloneWith(newFilters, orFilters, sorts, limitValue, offsetValue, pageValue,
                searchValue, afterCursor, beforeCursor);
    }

    public static class OrBuilder {
        private final List<FilterTuple> filters = new ArrayList<>();

        public OrBuilder where(String field, String op, Object value) {
            filters.add(new FilterTuple(field, op, value));
            return this;
        }

        public List<FilterTuple> getFilters() {
            return filters;
        }
    }

    public TableRef or(java.util.function.Consumer<OrBuilder> builderFn) {
        OrBuilder builder = new OrBuilder();
        builderFn.accept(builder);
        List<FilterTuple> newOrFilters = new ArrayList<>(orFilters);
        newOrFilters.addAll(builder.getFilters());
        return cloneWith(filters, newOrFilters, sorts, limitValue, offsetValue, pageValue,
                searchValue, afterCursor, beforeCursor);
    }

    public TableRef orderBy(String field) {
        return orderBy(field, "asc");
    }

    public TableRef orderBy(String field, String direction) {
        List<String[]> newSorts = new ArrayList<>(sorts);
        newSorts.add(new String[] { field, direction });
        return cloneWith(filters, orFilters, newSorts, limitValue, offsetValue, pageValue,
                searchValue, afterCursor, beforeCursor);
    }

    public TableRef limit(int n) {
        return cloneWith(filters, orFilters, sorts, n, offsetValue, pageValue,
                searchValue, afterCursor, beforeCursor);
    }

    public TableRef offset(int n) {
        return cloneWith(filters, orFilters, sorts, limitValue, n, pageValue,
                searchValue, afterCursor, beforeCursor);
    }

    public TableRef page(int n) {
        return cloneWith(filters, orFilters, sorts, limitValue, offsetValue, n,
                searchValue, afterCursor, beforeCursor);
    }

    public TableRef search(String query) {
        return cloneWith(filters, orFilters, sorts, limitValue, offsetValue, pageValue,
                query, afterCursor, beforeCursor);
    }

    public TableRef after(String cursor) {
        return cloneWith(filters, orFilters, sorts, limitValue, offsetValue, pageValue,
                searchValue, cursor, null);
    }

    public TableRef before(String cursor) {
        return cloneWith(filters, orFilters, sorts, limitValue, offsetValue, pageValue,
                searchValue, null, cursor);
    }

    // ─── CRUD ───

    @SuppressWarnings("unchecked")
    public ListResult getList() {
        Map<String, String> params = buildQueryParams();
        Map<String, Object> json;
        if (searchValue != null) {
            params.put("search", searchValue);
            json = (Map<String, Object>) coreGet("search", params, null);
        } else {
            json = (Map<String, Object>) coreGet("list", params, null);
        }
        return new ListResult(
                (List<Map<String, Object>>) json.getOrDefault("items", Collections.emptyList()),
                json.get("total") instanceof Number ? ((Number) json.get("total")).intValue() : null,
                json.get("page") instanceof Number ? ((Number) json.get("page")).intValue() : null,
                json.get("perPage") instanceof Number ? ((Number) json.get("perPage")).intValue() : null,
                json.get("hasMore") instanceof Boolean ? (Boolean) json.get("hasMore") : null,
                (String) json.get("cursor"));
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getOne(String id) {
        return doc(id).get();
    }

    /** Get the first record matching the current query conditions. Returns null if no match. */
    public Map<String, Object> getFirst() {
        ListResult result = limit(1).getList();
        return result.getItems().isEmpty() ? null : result.getItems().get(0);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> insert(Map<String, Object> record) {
        return (Map<String, Object>) coreInsert(record, Collections.emptyMap());
    }

    @SuppressWarnings("unchecked")
    public UpsertResult upsert(Map<String, Object> record) {
        return upsert(record, null);
    }

    @SuppressWarnings("unchecked")
    public UpsertResult upsert(Map<String, Object> record, String conflictTarget) {
        Map<String, String> query = new LinkedHashMap<>();
        query.put("upsert", "true");
        if (conflictTarget != null) {
            query.put("conflictTarget", conflictTarget);
        }
        Map<String, Object> json = (Map<String, Object>) coreInsert(record, query);
        return new UpsertResult(json, "inserted".equals(json.get("action")));
    }

    @SuppressWarnings("unchecked")
    public int count() {
        Map<String, String> params = buildQueryParams();
        Map<String, Object> json = (Map<String, Object>) coreGet("count", params, null);
        return json.get("total") instanceof Number ? ((Number) json.get("total")).intValue() : 0;
    }

    // ─── Batch Operations ───

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> insertMany(List<Map<String, Object>> records) {
        int chunkSize = 500;

        if (records.size() <= chunkSize) {
            Map<String, Object> json = (Map<String, Object>) coreBatch(
                    Map.of("inserts", records), Collections.emptyMap());
            return (List<Map<String, Object>>) json.getOrDefault("inserted", Collections.emptyList());
        }

        List<Map<String, Object>> allInserted = new ArrayList<>();
        for (int i = 0; i < records.size(); i += chunkSize) {
            List<Map<String, Object>> chunk = records.subList(i, Math.min(i + chunkSize, records.size()));
            Map<String, Object> json = (Map<String, Object>) coreBatch(
                    Map.of("inserts", chunk), Collections.emptyMap());
            List<Map<String, Object>> inserted = (List<Map<String, Object>>) json.getOrDefault("inserted",
                    Collections.emptyList());
            allInserted.addAll(inserted);
        }
        return allInserted;
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> upsertMany(List<Map<String, Object>> records) {
        return upsertMany(records, null);
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> upsertMany(List<Map<String, Object>> records, String conflictTarget) {
        int chunkSize = 500;
        Map<String, String> query = new LinkedHashMap<>();
        query.put("upsert", "true");
        if (conflictTarget != null) {
            query.put("conflictTarget", conflictTarget);
        }

        if (records.size() <= chunkSize) {
            Map<String, Object> json = (Map<String, Object>) coreBatch(
                    Map.of("inserts", records), query);
            return (List<Map<String, Object>>) json.getOrDefault("inserted", Collections.emptyList());
        }

        List<Map<String, Object>> allInserted = new ArrayList<>();
        for (int i = 0; i < records.size(); i += chunkSize) {
            List<Map<String, Object>> chunk = records.subList(i, Math.min(i + chunkSize, records.size()));
            Map<String, Object> json = (Map<String, Object>) coreBatch(
                    Map.of("inserts", chunk), query);
            List<Map<String, Object>> inserted = (List<Map<String, Object>>) json.getOrDefault("inserted",
                    Collections.emptyList());
            allInserted.addAll(inserted);
        }
        return allInserted;
    }

    public BatchResult updateMany(Map<String, Object> update) {
        if (filters.isEmpty())
            throw new IllegalArgumentException("updateMany requires at least one where() filter");
        return batchByFilter("update", update);
    }

    public BatchResult deleteMany() {
        if (filters.isEmpty())
            throw new IllegalArgumentException("deleteMany requires at least one where() filter");
        return batchByFilter("delete", null);
    }

    @SuppressWarnings("unchecked")
    private BatchResult batchByFilter(String action, Map<String, Object> update) {
        int maxIterations = 100;
        int totalProcessed = 0;
        int totalSucceeded = 0;
        List<Map<String, Object>> errors = new ArrayList<>();
        List<Object> filterJson = filters.stream().map(f -> (Object) f.toJson()).toList();

        for (int chunkIndex = 0; chunkIndex < maxIterations; chunkIndex++) {
            try {
                Map<String, Object> body = new HashMap<>();
                body.put("action", action);
                body.put("filter", filterJson);
                body.put("limit", 500);
                if (!orFilters.isEmpty()) {
                    body.put("orFilter", orFilters.stream().map(f -> (Object) f.toJson()).toList());
                }
                if ("update".equals(action) && update != null) {
                    body.put("update", update);
                }

                Map<String, Object> json = (Map<String, Object>) coreBatchByFilter(body, Collections.emptyMap());
                int processed = json.get("processed") instanceof Number
                        ? ((Number) json.get("processed")).intValue()
                        : 0;
                int succeeded = json.get("succeeded") instanceof Number
                        ? ((Number) json.get("succeeded")).intValue()
                        : 0;
                totalProcessed += processed;
                totalSucceeded += succeeded;

                if (processed == 0)
                    break;

                // For 'update', don't loop — updated records still match the filter,
                // so re-querying would process the same rows again (infinite loop).
                // Only 'delete' benefits from looping since deleted rows disappear.
                if ("update".equals(action))
                    break;
            } catch (Exception e) {
                errors.add(Map.of("chunkIndex", chunkIndex, "chunkSize", 500,
                        "error", e.getMessage() != null ? e.getMessage() : "Unknown error"));
                break;
            }
        }
        return new BatchResult(totalProcessed, totalSucceeded, errors);
    }

    // ─── Document Reference ───

    public DocRef doc(String id) {
        return new DocRef(core, namespace, instanceId, name, id, databaseLive);
    }

    // ─── DatabaseLive ─── 

    /**
     * Subscribe to table changes.
     * Only available when using client-side SDK (EdgeBase.client()).
     *
     * @param listener callback invoked for each change event
     * @return a Subscription that can be closed to unsubscribe
     */
    public DatabaseLiveClient.Subscription onSnapshot(java.util.function.Consumer<DbChange> listener) {
        if (databaseLive == null) {
            throw new UnsupportedOperationException(
                    "onSnapshot() is not available on the server SDK. Use EdgeBase.client() for database-live subscriptions.");
        }
        return databaseLive.subscribe(buildDatabaseLiveChannel(namespace, instanceId, name, null), change -> {
            if (matchesFilters(change.getRecord())) {
                listener.accept(change);
            }
        });
    }

    // ─── Internal ───

    private TableRef cloneWith(List<FilterTuple> filters, List<FilterTuple> orFilters, List<String[]> sorts,
            Integer limitValue, Integer offsetValue, Integer pageValue,
            String searchValue,
            String afterCursor, String beforeCursor) {
        return new TableRef(core, name, namespace, instanceId, databaseLive,
                filters, orFilters, sorts, limitValue,
                offsetValue, pageValue, searchValue,
                afterCursor, beforeCursor);
    }

    private Map<String, String> buildQueryParams() {
        boolean hasCursor = afterCursor != null || beforeCursor != null;
        boolean hasOffset = offsetValue != null || pageValue != null;
        if (hasCursor && hasOffset) {
            throw new IllegalArgumentException(
                    "Cannot use page()/offset() with after()/before() — choose offset or cursor pagination");
        }

        Map<String, String> params = new LinkedHashMap<>();
        if (!filters.isEmpty()) {
            List<Object> filterList = filters.stream().map(f -> (Object) f.toJson()).toList();
            params.put("filter", gson.toJson(filterList));
        }
        if (!orFilters.isEmpty()) {
            List<Object> orFilterList = orFilters.stream().map(f -> (Object) f.toJson()).toList();
            params.put("orFilter", gson.toJson(orFilterList));
        }
        if (!sorts.isEmpty()) {
            StringJoiner sj = new StringJoiner(",");
            for (String[] sort : sorts)
                sj.add(sort[0] + ":" + sort[1]);
            params.put("sort", sj.toString());
        }
        if (limitValue != null)
            params.put("limit", limitValue.toString());
        if (pageValue != null)
            params.put("page", pageValue.toString());
        if (offsetValue != null)
            params.put("offset", offsetValue.toString());
        if (afterCursor != null)
            params.put("after", afterCursor);
        if (beforeCursor != null)
            params.put("before", beforeCursor);
        return params;
    }

    private boolean matchesFilters(Map<String, Object> record) {
        if (record == null)
            return true;

        boolean andPass = filters.isEmpty() || filters.stream().allMatch(filter -> {
            Object fieldValue = record.get(filter.getField());
            return switch (filter.getOp()) {
                case "==" -> Objects.equals(fieldValue, filter.getValue());
                case "!=" -> !Objects.equals(fieldValue, filter.getValue());
                case ">" -> compareValues(fieldValue, filter.getValue()) > 0;
                case ">=" -> compareValues(fieldValue, filter.getValue()) >= 0;
                case "<" -> compareValues(fieldValue, filter.getValue()) < 0;
                case "<=" -> compareValues(fieldValue, filter.getValue()) <= 0;
                default -> true;
            };
        });
        if (!andPass)
            return false;

        if (!orFilters.isEmpty()) {
            boolean orPass = orFilters.stream().anyMatch(filter -> {
                Object fieldValue = record.get(filter.getField());
                return switch (filter.getOp()) {
                    case "==" -> Objects.equals(fieldValue, filter.getValue());
                    case "!=" -> !Objects.equals(fieldValue, filter.getValue());
                    case ">" -> compareValues(fieldValue, filter.getValue()) > 0;
                    case ">=" -> compareValues(fieldValue, filter.getValue()) >= 0;
                    case "<" -> compareValues(fieldValue, filter.getValue()) < 0;
                    case "<=" -> compareValues(fieldValue, filter.getValue()) <= 0;
                    default -> true;
                };
            });
            if (!orPass)
                return false;
        }

        return true;
    }

    @SuppressWarnings({ "unchecked", "rawtypes" })
    private int compareValues(Object a, Object b) {
        if (a == null && b == null)
            return 0;
        if (a == null)
            return -1;
        if (b == null)
            return 1;
        if (a instanceof Comparable && b instanceof Comparable) {
            return ((Comparable) a).compareTo(b);
        }
        return a.toString().compareTo(b.toString());
    }
}

File: packages/sdk/ruby/packages/core/lib/edgebase_core/table_ref.rb
# frozen_string_literal: true

require "json"

module EdgebaseCore
  # Query filter tuple.
  FilterTuple = Struct.new(:field_name, :op, :value) do
    def to_json_array
      [field_name, op, value]
    end
  end

  # Builder for OR conditions.
  class OrBuilder
    def initialize
      @filters = []
    end

    def where(field_name, op, value)
      @filters << FilterTuple.new(field_name, op, value)
      self
    end

    def get_filters
      @filters.dup
    end
  end

  # Collection query result — unified type for offset and cursor pagination.
  ListResult = Struct.new(:items, :total, :page, :per_page, :has_more, :cursor, keyword_init: true) do
    def initialize(items: [], total: nil, page: nil, per_page: nil, has_more: nil, cursor: nil)
      super(items: items, total: total, page: page, per_page: per_page, has_more: has_more, cursor: cursor)
    end
  end

  # Batch operation result.
  BatchResult = Struct.new(:total_processed, :total_succeeded, :errors, keyword_init: true)

  # Upsert operation result.
  UpsertResult = Struct.new(:record, :inserted, keyword_init: true)

  # DatabaseLive database change event.
  DbChange = Struct.new(:event, :table, :id, :record, :old_record, keyword_init: true) do
    def self.from_json(data)
      new(
        event: data["event"] || "",
        table: data["table"] || "",
        id: data["id"],
        record: data["record"],
        old_record: data["oldRecord"]
      )
    end
  end

  def self.build_database_live_channel(namespace, table, instance_id = nil, doc_id = nil)
    base = instance_id ? "dblive:#{namespace}:#{instance_id}:#{table}" : "dblive:#{namespace}:#{table}"
    doc_id ? "#{base}:#{doc_id}" : base
  end

  # ── Core dispatch helpers ──────────────────────────────────────────────────

  module CoreDispatch
    module_function

    def core_get(core, method, namespace, instance_id, table, doc_id: nil, query: nil)
      if instance_id
        case method
        when "list"   then core.db_list_records(namespace, instance_id, table, query: query)
        when "get"    then core.db_get_record(namespace, instance_id, table, doc_id, query: query)
        when "count"  then core.db_count_records(namespace, instance_id, table, query: query)
        when "search" then core.db_search_records(namespace, instance_id, table, query: query)
        end
      else
        case method
        when "list"   then core.db_single_list_records(namespace, table, query: query)
        when "get"    then core.db_single_get_record(namespace, table, doc_id, query: query)
        when "count"  then core.db_single_count_records(namespace, table, query: query)
        when "search" then core.db_single_search_records(namespace, table, query: query)
        end
      end
    end

    def core_insert(core, namespace, instance_id, table, body, query = nil)
      if instance_id
        core.db_insert_record(namespace, instance_id, table, body, query: query)
      else
        core.db_single_insert_record(namespace, table, body, query: query)
      end
    end

    def core_update(core, namespace, instance_id, table, doc_id, body)
      if instance_id
        core.db_update_record(namespace, instance_id, table, doc_id, body)
      else
        core.db_single_update_record(namespace, table, doc_id, body)
      end
    end

    def core_delete(core, namespace, instance_id, table, doc_id)
      if instance_id
        core.db_delete_record(namespace, instance_id, table, doc_id)
      else
        core.db_single_delete_record(namespace, table, doc_id)
      end
    end

    def core_batch(core, namespace, instance_id, table, body, query = nil)
      if instance_id
        core.db_batch_records(namespace, instance_id, table, body, query: query)
      else
        core.db_single_batch_records(namespace, table, body, query: query)
      end
    end

    def core_batch_by_filter(core, namespace, instance_id, table, body)
      if instance_id
        core.db_batch_by_filter(namespace, instance_id, table, body, query: nil)
      else
        core.db_single_batch_by_filter(namespace, table, body, query: nil)
      end
    end
  end

  # ── TableRef ───────────────────────────────────────────────────────────────

  # Immutable table reference with query builder.
  #
  # All chaining methods return a new instance — safe for reference sharing.
  # All HTTP calls delegate to Generated Core (no hardcoded paths).
  #
  #   posts = client.db("shared").table("posts")
  #   result = posts.where("status", "==", "published")
  #                  .order_by("createdAt", "desc")
  #                  .limit(20)
  #                  .get_list
  class TableRef
    attr_reader :_name, :_namespace, :_instance_id, :_filters, :_or_filters,
                :_sorts, :_limit, :_offset, :_page, :_search, :_after, :_before

    def initialize(core, name, database_live: nil, namespace: "shared", instance_id: nil,
                   filters: nil, or_filters: nil, sorts: nil,
                   limit_value: nil, offset_value: nil, page_value: nil,
                   search_value: nil, after_value: nil, before_value: nil)
      @core = core
      @_name = name
      @database_live = database_live
      @_namespace = namespace
      @_instance_id = instance_id
      @_filters = filters || []
      @_or_filters = or_filters || []
      @_sorts = sorts || []
      @_limit = limit_value
      @_offset = offset_value
      @_page = page_value
      @_search = search_value
      @_after = after_value
      @_before = before_value
    end

    # ── Query Builder (immutable) ──────────────────────────────────────────

    def where(field_name, op, value)
      clone_with(filters: [*@_filters, FilterTuple.new(field_name, op, value)])
    end

    def or_(&block)
      builder = OrBuilder.new
      block.call(builder)
      clone_with(or_filters: [*@_or_filters, *builder.get_filters])
    end

    def order_by(field_name, direction = "asc")
      clone_with(sorts: [*@_sorts, [field_name, direction]])
    end

    def limit(n)
      clone_with(limit_value: n)
    end

    def offset(n)
      clone_with(offset_value: n)
    end

    # Set page number for offset pagination (1-based).
    def page(n)
      clone_with(page_value: n)
    end

    def search(query)
      clone_with(search_value: query)
    end

    # Set cursor for forward pagination. Mutually exclusive with offset().
    def after(cursor)
      clone_with(after_value: cursor, before_value: nil)
    end

    # Set cursor for backward pagination. Mutually exclusive with offset().
    def before(cursor)
      clone_with(before_value: cursor, after_value: nil)
    end

    # ── CRUD ───────────────────────────────────────────────────────────────

    def get_list
      params = build_query_params
      if @_search
        params["search"] = @_search
        data = CoreDispatch.core_get(@core, "search", @_namespace, @_instance_id, @_name, query: params)
      else
        data = CoreDispatch.core_get(@core, "list", @_namespace, @_instance_id, @_name, query: params)
      end
      return ListResult.new(items: []) unless data.is_a?(Hash)

      ListResult.new(
        items: data["items"] || [],
        total: data["total"],
        page: data["page"],
        per_page: data["perPage"],
        has_more: data["hasMore"],
        cursor: data["cursor"]
      )
    end

    # Get a single record by ID.
    def get_one(doc_id)
      CoreDispatch.core_get(
        @core, "get", @_namespace, @_instance_id, @_name,
        doc_id: doc_id, query: {}
      )
    end

    def insert(record)
      CoreDispatch.core_insert(@core, @_namespace, @_instance_id, @_name, record)
    end

    def upsert(record, conflict_target: nil)
      query = { "upsert" => "true" }
      query["conflictTarget"] = conflict_target if conflict_target
      data = CoreDispatch.core_insert(@core, @_namespace, @_instance_id, @_name, record, query)
      UpsertResult.new(
        record: data.is_a?(Hash) ? data : {},
        inserted: data.is_a?(Hash) && data["action"] == "inserted"
      )
    end

    def count
      params = build_query_params
      data = CoreDispatch.core_get(@core, "count", @_namespace, @_instance_id, @_name, query: params)
      data.is_a?(Hash) ? (data["total"] || 0) : 0
    end

    # Get the first record matching the current query conditions.
    def get_first
      result = self.limit(1).get_list
      result.items.first
    end

    def update(doc_id, data)
      doc(doc_id).update(data)
    end

    def delete(doc_id)
      doc(doc_id).delete
    end

    def sql(query, params = [])
      body = {
        "namespace" => @_namespace,
        "sql" => query,
        "params" => params
      }
      body["id"] = @_instance_id unless @_instance_id.nil?
      result = @core.http.post("/sql", body)
      result.is_a?(Hash) ? (result["items"] || []) : []
    end

    # ── Batch ──────────────────────────────────────────────────────────────

    # Create multiple records. Auto-chunks into 500-item batches.
    def insert_many(records)
      chunk_size = 500
      if records.length <= chunk_size
        data = CoreDispatch.core_batch(@core, @_namespace, @_instance_id, @_name, { "inserts" => records })
        return data.is_a?(Hash) ? (data["inserted"] || []) : []
      end

      all_inserted = []
      records.each_slice(chunk_size) do |chunk|
        data = CoreDispatch.core_batch(@core, @_namespace, @_instance_id, @_name, { "inserts" => chunk })
        all_inserted.concat(data["inserted"] || []) if data.is_a?(Hash)
      end
      all_inserted
    end

    # Upsert multiple records. Auto-chunks 500 items.
    def upsert_many(records, conflict_target: nil)
      chunk_size = 500
      query = { "upsert" => "true" }
      query["conflictTarget"] = conflict_target if conflict_target

      if records.length <= chunk_size
        data = CoreDispatch.core_batch(@core, @_namespace, @_instance_id, @_name, { "inserts" => records }, query)
        return data.is_a?(Hash) ? (data["inserted"] || []) : []
      end

      all_inserted = []
      records.each_slice(chunk_size) do |chunk|
        data = CoreDispatch.core_batch(@core, @_namespace, @_instance_id, @_name, { "inserts" => chunk }, query)
        all_inserted.concat(data["inserted"] || []) if data.is_a?(Hash)
      end
      all_inserted
    end

    # Update records matching query builder filters.
    def update_many(update)
      raise ArgumentError, "update_many requires at least one where() filter" if @_filters.empty?
      batch_by_filter("update", update)
    end

    # Delete records matching query builder filters.
    def delete_many
      raise ArgumentError, "delete_many requires at least one where() filter" if @_filters.empty?
      batch_by_filter("delete", nil)
    end

    # ── Doc ─────────────────────────────────────────────────────────────────

    def doc(doc_id)
      DocRef.new(
        @core, @_name, doc_id, @database_live,
        namespace: @_namespace, instance_id: @_instance_id
      )
    end

    # ── DatabaseLive ────────────────────────────────────────────────────────────

    def on_snapshot(&callback)
      raise "DatabaseLive not available" unless @database_live
      @database_live.subscribe_callback(
        EdgebaseCore.build_database_live_channel(@_namespace, @_name, @_instance_id),
        callback
      )
    end

    # ── Internal ────────────────────────────────────────────────────────────

    def build_query_params
      has_cursor = !@_after.nil? || !@_before.nil?
      has_offset = !@_offset.nil? || !@_page.nil?
      if has_cursor && has_offset
        raise ArgumentError,
              "Cannot use page()/offset() with after()/before() — choose offset or cursor pagination"
      end

      params = {}
      unless @_filters.empty?
        params["filter"] = JSON.generate(@_filters.map(&:to_json_array))
      end
      unless @_or_filters.empty?
        params["orFilter"] = JSON.generate(@_or_filters.map(&:to_json_array))
      end
      unless @_sorts.empty?
        params["sort"] = @_sorts.map { |f, d| "#{f}:#{d}" }.join(",")
      end
      params["limit"] = @_limit.to_s unless @_limit.nil?
      params["page"] = @_page.to_s unless @_page.nil?
      params["offset"] = @_offset.to_s unless @_offset.nil?
      params["after"] = @_after unless @_after.nil?
      params["before"] = @_before unless @_before.nil?
      params
    end

    private

    def clone_with(**kwargs)
      TableRef.new(
        @core, @_name,
        database_live: @database_live,
        namespace: @_namespace,
        instance_id: @_instance_id,
        filters: kwargs.fetch(:filters, @_filters),
        or_filters: kwargs.fetch(:or_filters, @_or_filters),
        sorts: kwargs.fetch(:sorts, @_sorts),
        limit_value: kwargs.fetch(:limit_value, @_limit),
        offset_value: kwargs.fetch(:offset_value, @_offset),
        page_value: kwargs.fetch(:page_value, @_page),
        search_value: kwargs.fetch(:search_value, @_search),
        after_value: kwargs.fetch(:after_value, @_after),
        before_value: kwargs.fetch(:before_value, @_before)
      )
    end

    def batch_by_filter(action, update)
      max_iterations = 100
      total_processed = 0
      total_succeeded = 0
      errors = []
      filter_json = @_filters.map(&:to_json_array)

      max_iterations.times do |chunk_index|
        body = {
          "action" => action,
          "filter" => filter_json,
          "limit" => 500
        }
        body["orFilter"] = @_or_filters.map(&:to_json_array) unless @_or_filters.empty?
        body["update"] = update if action == "update" && update

        begin
          data = CoreDispatch.core_batch_by_filter(@core, @_namespace, @_instance_id, @_name, body)
          processed = data.is_a?(Hash) ? (data["processed"] || 0) : 0
          succeeded = data.is_a?(Hash) ? (data["succeeded"] || 0) : 0
          total_processed += processed
          total_succeeded += succeeded

          break if processed == 0
          # For 'update', don't loop — updated records still match the filter
          break if action == "update"
        rescue StandardError => e
          errors << { "chunkIndex" => chunk_index, "chunkSize" => 500, "error" => e.to_s }
          break
        end
      end

      BatchResult.new(
        total_processed: total_processed,
        total_succeeded: total_succeeded,
        errors: errors
      )
    end
  end

  # ── DocRef ───────────────────────────────────────────────────────────────────

  # Document reference for single-document operations.
  class DocRef
    attr_reader :table_name, :id

    def initialize(core, table_name, doc_id, database_live = nil, namespace: "shared", instance_id: nil)
      @core = core
      @table_name = table_name
      @id = doc_id
      @database_live = database_live
      @_namespace = namespace
      @_instance_id = instance_id
    end

    def get
      CoreDispatch.core_get(
        @core, "get", @_namespace, @_instance_id, @table_name,
        doc_id: @id, query: {}
      )
    end

    def update(data)
      CoreDispatch.core_update(@core, @_namespace, @_instance_id, @table_name, @id, data)
    end

    def delete
      CoreDispatch.core_delete(@core, @_namespace, @_instance_id, @table_name, @id)
    end

    def on_snapshot(&callback)
      raise "DatabaseLive not available" unless @database_live
      @database_live.subscribe_callback(
        EdgebaseCore.build_database_live_channel(@_namespace, @table_name, @_instance_id, @id),
        callback
      )
    end
  end
end
